backend/src/user/user.service.ts

```diff
@@ -1,33 +1,39 @@
-import { Injectable } from '@nestjs/common';
+import { ForbiddenException, Injectable } from '@nestjs/common';
 import { PrismaService } from 'src/prisma/prisma.service';
+import { User } from '@prisma/client';
+import { Prisma } from '@prisma/client';
 
 import { loginDto, signUpDto } from './dto/user.dto';
 @Injectable()
 export class UserService {
   constructor(private prisma: PrismaService) {}
 
-  async signUp(dto: signUpDto) {
-    // エラー処理は書いてない
+  async signUp(dto: signUpDto): Promise<User> {
     console.log(dto);
-    const user = this.prisma.user.create({
-      data: {
-        email: dto.email,
-        nickname: dto.nickname,
-        // 今後ハッシュ化
-        hashedPassword: dto.hashedPassword,
-      },
-    });
-    user
-      .then((e) => {
-        console.log(e);
-      })
-      .catch((e) => {
-        // emailが被った時のエラーは'P2002'が帰ってくる
-        console.log(e);
+    try {
+      const user = await this.prisma.user.create({
+        data: {
+          email: dto.email,
+          nickname: dto.nickname,
+          // 今後ハッシュ化
+          hashedPassword: dto.hashedPassword,
+        },
       });
+      return user;
+    } catch (e) {
+      console.log(e);
+      // email,nicknameが被った時のエラーは'P2002'が帰ってくる
+      if (e instanceof Prisma.PrismaClientKnownRequestError) {
+        if (e.code === 'P2002') {
+          throw new ForbiddenException('Email or nickname is already taken');
+        }
+      }
+      // 500 internal server err
+      throw e;
+    }
   }
 
-  async login(dto: loginDto) {
+  async login(dto: loginDto): Promise<User> {
```

Comment:
signUp の方が try catch で囲まれてるけど こっちはいらないの？

Reply:
[findUnique return type](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#return-type)

> - Error
  > If rejectOnNotFound is true, findUnique throws an error (NotFoundError by default, [customizable globally](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#rejectonnotfound)) instead of returning null.

非推奨みたいだからデフォルトはオフっぽいな
default falseってどこにも書いてなさそうだけど・・・・

Reply:
もしかして全てのapiについて、あらゆる例外があり得るのか？
https://www.prisma.io/docs/reference/api-reference/error-reference
noExceptみたいなのないのかな、全部try catchする必要あり？

Reply:
そこの例外は例外フィルターでキャッチできてまする

Reply:
すげえ